app/assets/javascripts/admin/subscriptions/controllers/details_controller.js.coffee

```diff
@@ -1,38 +1,43 @@
-angular.module("admin.subscriptions").controller "DetailsController", ($scope, $http, CreditCardResource, StatusMessage) ->
+angular.module("admin.subscriptions").controller "DetailsController", ($scope, $http, CustomerResource, StatusMessage) ->
   $scope.cardRequired = false
 
   $scope.registerNextCallback 'details', ->
     $scope.subscription_form.$submitted = true
-    if $scope.subscription_details_form.$valid
-      $scope.subscription_form.$setPristine()
-      StatusMessage.clear()
-      $scope.setView('address')
-    else
-      StatusMessage.display 'failure', t('admin.subscriptions.details.invalid_error')
+    return unless $scope.validate()
+    $scope.subscription_form.$setPristine()
+    StatusMessage.clear()
+    $scope.setView('address')
 
   $scope.$watch "subscription.customer_id", (newValue, oldValue) ->
     return if !newValue?
-    $scope.loadAddresses(newValue) unless $scope.subscription.id?
-    $scope.loadCreditCards(newValue)
+    $scope.loadCustomer(newValue) unless $scope.subscription.id?
 
   $scope.$watch "subscription.payment_method_id", (newValue, oldValue) ->
     return if !newValue?
     paymentMethod = ($scope.paymentMethods.filter (pm) -> pm.id == newValue)[0]
     return unless paymentMethod?
-    if paymentMethod.type == "Spree::Gateway::StripeConnect"
-      $scope.cardRequired = true
-    else
-      $scope.cardRequired = false
-      $scope.subscription.credit_card_id = null
+    $scope.cardRequired = (paymentMethod.type == "Spree::Gateway::StripeConnect")
+    $scope.loadCustomer() if $scope.cardRequired && !$scope.customer
 
-  $scope.loadAddresses = (customer_id) ->
-    $http.get("/admin/customers/#{customer_id}/addresses")
-    .success (response) =>
-      delete response.bill_address.id
-      delete response.ship_address.id
-      angular.extend($scope.subscription.bill_address, response.bill_address)
-      angular.extend($scope.subscription.ship_address, response.ship_address)
-      $scope.shipAddressFromBilling() unless response.ship_address.address1?
+  $scope.loadCustomer = ->
+    params = { id: $scope.subscription.customer_id }
+    params.ams_prefix = 'subscription' unless $scope.subscription.id
+    $scope.customer = CustomerResource.get params, (response) ->
+      for address in ['bill_address','ship_address']
```

Comment:
I believe [forEach](https://docs.angularjs.org/api/ng/function/angular.forEach) is better suited for this. It feels more natural in a functional language like JS.

Reply:
Isn't there forEach as well? Goddamn hipsters... :trollface: